glean/src/core/upload/ping_request.ts

```diff
@@ -0,0 +1,51 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { gzipSync, strToU8 } from "fflate";
+import { PingBodyOverflowError } from "./ping_body_overflow_error.js";
+
+/**
+ * Represents a payload to be transmitted by an uploading mechanism.
+ */
+export default class PingRequest<BodyType extends string | Uint8Array> {
+  constructor (
+    readonly identifier: string,
+    readonly headers: Record<string, string>,
+    readonly payload: BodyType,
+    readonly maxBodySize: number
+  ) {}
+
+  asCompressedPayload(): PingRequest<string | Uint8Array> {
+    // If this is already gzipped, do nothing.
+    if (this.headers["Content-Encoding"] == "gzip") {
+      return this;
+    }
+
+    const stringifiedBody = JSON.stringify(this.payload);
+    // We prefer using `strToU8` instead of TextEncoder directly,
+    // because it will polyfill TextEncoder if it's not present in the environment.
+    // Environments that don't provide TextEncoder are IE and most importantly QML.
```

Comment:
```suggestion
    // Environments that don't provide TextEncoder are IE and most importantly QML.
```

While you are here, probably worth changing this.
We don't have QML support anymore.